crates/server/src/server/server_future.rs

```diff
@@ -70,7 +70,7 @@ impl<T: RequestHandler> ServerFuture<T> {
                     let message = match message {
                         Err(e) => {
                             warn!("error receiving message on udp_socket: {}", e);
-                            continue;
+                            break;
```

Comment:
hm, we should think about this a bit more, that is, we probably need an outer loop to recreate the UdpSocket. As the code stands here, we'll break out of the spawn in tokio and stop listening entirely on socket. I think it means reworking the interface on this function though, to move the socket instantiation into the `register_socket` (i.e. instead of taking a `Socket`, it would take an `IpAddr`.

Reply:
Indeed. I was trying to think of ways to perhaps allow this code to mutably borrow the socket instead, which would give the calling code an opportunity to address any errors we'd return here. Maybe that means recreating the socket, or maybe it means removing it from the socket set and retrying (not possible yet, but a possibility in the future).

Either way, the outside code has no way of knowing anything is even wrong without us breaking out, and the code as it stands doesn't have any meaningful recovery mechanisms. So for now, I think breaking early is the only remaining option.

Reply:
I also have a local fork that uses JoinSet from tokio to ensure all resolver tasks spawned are aborted when the server is dropped, but unfortunately that requires "unstable" tokio to build. I can post it here, if you'd like to take a look. It does seem to allow all the owned bound sockets to drop and close before the await returns, and it also aborts any in-flight requests.

Reply:
I think you're right that the documentation is't totally accurate. The goal was initially to ensure that the server would shutdown if any of the tasks finished (which is basically unexpected). We'd need to add a signal to all of those tasks that shutdown has been initiated and then give them a timeout before shutting down.

Maybe this implementation you've put forward is fine, and restarting the process would via some monitor process (which I'd expect anyone doing this seriously would use). Feel free to change the documentation on that method if you want. It should read "await all tasks and exit if any return early" or something along those lines.